pkg/cluster/update_controller.go

```diff
@@ -0,0 +1,290 @@
+	// provided version.
+	UpdateToVersion(*components.Version) (bool, error)
+	// Priority is the priority level for this component.
+	Priority() int
+	// Version of the component.
+	Version() (*components.Version, error)
+}
+
+// NewClusterUpdater returns a ClusterUpdater struct with defaults.
+func NewClusterUpdater(client clientset.Interface, uc unversioned.Interface) (*UpdateController, error) {
+	l, err := labels.Parse(clusterManagedLabel)
+	if err != nil {
+		return nil, err
+	}
+	mlo := api.ListOptions{LabelSelector: l}
+	nodeStore, nodeController := framework.NewInformer(
+		&cache.ListWatch{
+			ListFunc: func(_ api.ListOptions) (runtime.Object, error) {
+				return uc.Nodes().List(mlo)
+			},
+			WatchFunc: func(_ api.ListOptions) (watch.Interface, error) {
+				return uc.Nodes().Watch(mlo)
+			},
+		},
+		&api.Node{},
+		30*time.Minute,
+		framework.ResourceEventHandlerFuncs{},
+	)
+	daemonSetStore, daemonSetController := framework.NewInformer(
+		&cache.ListWatch{
+			ListFunc: func(_ api.ListOptions) (runtime.Object, error) {
+				return uc.Extensions().DaemonSets(api.NamespaceSystem).List(mlo)
+			},
+			WatchFunc: func(_ api.ListOptions) (watch.Interface, error) {
+				return uc.Extensions().DaemonSets(api.NamespaceSystem).Watch(mlo)
+			},
+		},
+		&extensions.DaemonSet{},
+		30*time.Minute,
+		framework.ResourceEventHandlerFuncs{},
+	)
+	deploymentStore, deploymentController := framework.NewInformer(
+		&cache.ListWatch{
+			ListFunc: func(_ api.ListOptions) (runtime.Object, error) {
+				return uc.Extensions().Deployments(api.NamespaceSystem).List(mlo)
+			},
+			WatchFunc: func(_ api.ListOptions) (watch.Interface, error) {
+				return uc.Extensions().Deployments(api.NamespaceSystem).Watch(mlo)
+			},
+		},
+		&extensions.Deployment{},
+		30*time.Minute,
+		framework.ResourceEventHandlerFuncs{},
+	)
+	podStore, podController := framework.NewInformer(
+		&cache.ListWatch{
+			ListFunc: func(lo api.ListOptions) (runtime.Object, error) {
+				return uc.Pods(api.NamespaceSystem).List(lo)
+			},
+			WatchFunc: func(lo api.ListOptions) (watch.Interface, error) {
+				return uc.Pods(api.NamespaceSystem).Watch(lo)
+			},
+		},
+		&api.Pod{},
+		30*time.Minute,
+		framework.ResourceEventHandlerFuncs{},
+	)
+
+	go nodeController.Run(wait.NeverStop)
+	go daemonSetController.Run(wait.NeverStop)
+	go deploymentController.Run(wait.NeverStop)
+	go podController.Run(wait.NeverStop)
+
+	return &UpdateController{
+		Client:                    client,
+		UnversionedClient:         uc,
+		GetAllManagedComponentsFn: DefaultGetAllManagedComponentsFn,
+		nodes:       cache.StoreToNodeLister{nodeStore},
+		deployments: cache.StoreToDeploymentLister{deploymentStore},
+		daemonSets:  cache.StoreToDaemonSetLister{daemonSetStore},
+		pods:        components.StoreToPodLister{podStore},
+	}, nil
+}
+
+// UpdateToVersion will update the cluster to the given version.
+func (cu *UpdateController) UpdateToVersion(v *components.Version) error {
```

Comment:
It looks like we are only re-calling this on configMap re-sync (or change). However, because we are exiting this process after each component update - we should re-trigger this more often.

Might make sense to have 2 retry periods on the caller. If we're in the process of an update, we should re-enter this quickly (if not right away). If no update was made, or an error -- we should have a longer retry period.